syntax = "proto3";

message LogEntry {
  int32 term = 1;
  string operation = 2;
}

message AppendEntryRequest {
  int32 term = 1;
  int32 leader_id = 2;
  int32 prev_log_index = 3;
  int32 prev_log_term = 4;
  repeated LogEntry entries = 5;
  int32 leader_commit = 6;
}

message AppendEntryResponse {
  int32 term = 1;
  bool success = 2;
}

message RequestVoteRequest {
  int32 term = 1;
  int32 candidate_id = 2;
  int32 last_log_index = 3;
  int32 last_log_term = 4;
}

message RequestVoteResponse {
  int32 term = 1;
  bool vote_granted = 2;
}

service RaftNode {
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
  rpc AppendEntry(AppendEntryRequest) returns (AppendEntryResponse);
}
